File: en/codes/kotlin/chapter_searching/hashing_search.kt
/**
 * File: hashing_search.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_searching

import utils.ListNode

/* Hash search (array) */
fun hashingSearchArray(map: Map<Int?, Int>, target: Int): Int {
    // Hash table key: Target element, _val: Index
    // If the hash table does not contain this key, return -1
    return map.getOrDefault(target, -1)
}

/* Hash search (linked list) */
fun hashingSearchLinkedList(map: Map<Int?, ListNode?>, target: Int): ListNode? {
    // Hash table key: Target node value, _val: Node object
    // If the key is not in the hash table, return null
    return map.getOrDefault(target, null)
}

/* Driver Code */
fun main() {
    val target = 3

    /* Hash search (array) */
    val nums = intArrayOf(1, 5, 3, 2, 4, 7, 5, 9, 10, 8)
    // Initialize hash table
    val map = HashMap<Int?, Int>()
    for (i in nums.indices) {
        map[nums[i]] = i // key: Element, _val: Index
    }
    val index = hashingSearchArray(map, target)
    println("Index of target element 3 = $index")

    /* Hash search (linked list) */
    var head = ListNode.arrToLinkedList(nums)
    // Initialize hash table
    val map1 = HashMap<Int?, ListNode?>()
    while (head != null) {
        map1[head._val] = head // key: Node value, _val: Node
        head = head.next
    }
    val node = hashingSearchLinkedList(map1, target)
    println("Corresponding node object for target node value 3 = $node")
}

File: en/codes/rust/chapter_searching/hashing_search.rs
/*
 * File: hashing_search.rs
 * Created Time: 2023-07-09
 * Author: codingonion (coderonion@gmail.com)
 */

include!("../include/include.rs");

use list_node::ListNode;
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

/* Hash search (array) */
fn hashing_search_array<'a>(map: &'a HashMap<i32, usize>, target: i32) -> Option<&'a usize> {
    // Hash table's key: target element, value: index
    // If the hash table does not contain this key, return None
    map.get(&target)
}

/* Hash search (linked list) */
fn hashing_search_linked_list(
    map: &HashMap<i32, Rc<RefCell<ListNode<i32>>>>,
    target: i32,
) -> Option<&Rc<RefCell<ListNode<i32>>>> {
    // Hash table key: target node value, value: node object
    // If the hash table does not contain this key, return None
    map.get(&target)
}

/* Driver Code */
pub fn main() {
    let target = 3;

    /* Hash search (array) */
    let nums = [1, 5, 3, 2, 4, 7, 5, 9, 10, 8];
    // Initialize hash table
    let mut map = HashMap::new();
    for (i, num) in nums.iter().enumerate() {
        map.insert(*num, i); // key: element, value: index
    }
    let index = hashing_search_array(&map, target);
    println!("The index of target element 3 is {}", index.unwrap());

    /* Hash search (linked list) */
    let head = ListNode::arr_to_linked_list(&nums);
    // Initialize hash table
    // let mut map1 = HashMap::new();
    let map1 = ListNode::linked_list_to_hashmap(head);
    let node = hashing_search_linked_list(&map1, target);
    println!("The corresponding node object for target value 3 is {:?}", node);
}
